lib/spec/router.sw-spec.js

```diff
@@ -4,43 +4,186 @@ importScripts('/base/lib/router.js');
+      router.methods.forEach(function (verb) {
+        expect(router[verb]).to.be.defined;
+      });
+    });
+
+    it('should wrap add()', function() {
+      var handler = function() {};
+      sinon.spy(router, 'add');
+      router.get('/', handler);
+      expect(router.add.calledWith('get', '/', handler)).to.be.true;
+    });
   });
 
-  it('should recover all the middlewares attached to a route and a method',
-  function () {
-    var router = new Router();
-    var mw1 = function () {};
-    var mw2 = function () {};
-    var mw3 = function () {};
-    router.get('/a', mw1);
-    router.get('/a', mw2);
-    router.post('/a', mw3);
-    router.get('/b', mw3);
-
-    expect(router.match('get', '/a')).to.deep.equal([mw1, mw2]);
-    expect(router.match('post', '/a')).to.deep.equal([mw3]);
-    expect(router.match('get', '/b')).to.deep.equal([mw3]);
+  describe('proxyMethods()', function() {
+    var object;
+    var handler = function() {};
+
+    beforeEach(function() {
+      object = {};
+    });
+
+    it('should proxy the methods to an object', function () {
+      router.proxyMethods(object);
+      router.methods.forEach(function (verb) {
+        expect(object[verb]).to.be.defined;
+      });
+    });
+
+    it('should wrap add()', function() {
+      sinon.spy(router, 'add');
+      router.proxyMethods(object);
+      object.get('/', handler);
+      expect(router.add.calledWith('get', '/', handler)).to.be.true;
+    });
+
+    // Currently failing, due to a testability issue: the handler gets binded,
+    // which prevents to test the equality of the handlers.
+    // If you guys have any idea on how to improve that.
+    it('should support an object as a middleware', function() {
+      var cb = sinon.spy(router, 'add');
+      var handler = function() {};
+      var handlerWithOnFetch = {
+        onFetch: handler
+      };
+      router.proxyMethods(object);
+      object.get('/', handlerWithOnFetch);
+      var bindedHandler = handlerWithOnFetch.onFetch.bind(handlerWithOnFetch);
+      expect(router.add.calledWith('get', '/', bindedHandler)).to.be.true;
+    });
+
+    it('should throw an error if the object does not have onFetch()', function() {
+      var invalidHandler = {};
+      router.proxyMethods(object);
+      expect(function() { object.get('/', invalidHandler); }).to.throw(Error, 'This middleware cannot handle fetch request')
+    });
+  });
+
+  describe('match()', function() {
+    var defaultMiddleware = function() {};
+
+    it('should match a single middleware', function() {
+      router.get('/', defaultMiddleware);
+      expect(router.match('get', '/')).to.deep.equal([defaultMiddleware]);
+    });
+
+    it('should match multiple middlewares', function() {
+      var mw2 = function () {};
+      router.get('/', defaultMiddleware);
+      router.get('/', mw2);
+      expect(router.match('get', '/')).to.deep.equal([defaultMiddleware, mw2]);
```

Comment:
Salva made a great point here, is not just about the length but about order of mw to be executed, as they will modify the behavior of the response.

@JohanLorenzo the router does implicit do this since it matches the different routes in the specific order that we declare them.